### ISSUE TITLE
Xamarin.forms iOS app signed with invalid signature when deploying in debug on device

### ISSUE THREAD

**Issue by Bishokouts** [xamarin/xamarin-macios#18772]
### Steps to Reproduce

1. Set Apple signing to automatic with correct team for app bundle id
2. Build and run to actual ios device

### Expected Behavior

App built and signed and debugging on ios device.

### Actual Behavior

App builds, but during deploy the following error is output in the deploy window and the app is not installed:

ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.XhLS8d/extracted/App.iOS.app : 0xe8008014 (The executable contains an invalid signature.)
error MT1006: Could not install the application '[my app path and name here]' on the device '[device name here]': AMDeviceSecureInstallApplicationBundle returned:   ##0xe8008014.

I already tried solution mentioned in #16190. I had "EnableCodeSigning" key in my iOS project file with "true" value but I still get error.

```xml
<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|iPhone' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>portable</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\iPhone\Debug</OutputPath>
    <DefineConstants>DEBUG</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <MtouchArch>ARM64</MtouchArch>
    <CodesignKey>Apple Development: </CodesignKey>
    <MtouchDebug>true</MtouchDebug>
    <MtouchLink>None</MtouchLink>
    <MtouchInterpreter>-all</MtouchInterpreter>
    <CodesignProvision>VS: XXX.XXXXX.XXXXXXXXX Development</CodesignProvision>
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
    <RunWithWarnings>true</RunWithWarnings>
    <CodesignExtraArgs></CodesignExtraArgs>
    <MtouchUseThumb></MtouchUseThumb>
    <IpaPackageName></IpaPackageName>
    <CrashReportingApiKey></CrashReportingApiKey>
    <EnableCodeSigning>True</EnableCodeSigning>
  </PropertyGroup>
```

### Environment

<details>
<summary>Version information</summary>

```
Xcode Version 14.3.1 (14E300c)

Visual Studio Community 2022 for Mac
Version 17.6.3 (build 421)
Installation de l’identificateur global unique (UUID) : 50ee650b-8d10-4b90-bf39-892d459d564f

Runtime
.NET 7.0.3 (64-bit)
Architecture: X64
Microsoft.macOS.Sdk 13.1.1007; git-rev-head:8afca776a0a96613dfb7200e0917bb57f9ed5583; git-branch:release/7.0.1xx-xcode14.2

Roslyn (Language Service)
4.6.0-3.23180.6+99e956e42697a6dd886d1e12478ea2b27cceacfa

NuGet
Version : 6.4.0.117

Kit de développement logiciel (SDK) .NET (x64)
SDK : /usr/local/share/dotnet/sdk/7.0.307/Sdks
Versions du SDK :
	7.0.307
	7.0.306
	7.0.304
	7.0.302
	7.0.203
	7.0.101
	7.0.100
	6.0.413
	6.0.412
	6.0.410
	6.0.408
	6.0.404
	6.0.403
	6.0.106
	6.0.105
	6.0.104
	6.0.103
	6.0.102
	6.0.101
	6.0.100
	5.0.408
	5.0.407
	5.0.406
	5.0.405
	5.0.404
	5.0.403
	5.0.402
	5.0.401
	5.0.400
	5.0.302
	5.0.301
	3.1.426
	3.1.425
	3.1.420
	3.1.419
	3.1.418
	3.1.417
	3.1.416
	3.1.415
	3.1.414
	3.1.413
	3.1.412
	3.1.411
	3.1.410
SDK MSBuild : /Applications/Visual Studio.app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

Runtime .NET (x64)
Runtime : /usr/local/share/dotnet/dotnet
Versions du runtime :
	7.0.10
	7.0.9
	7.0.7
	7.0.5
	7.0.1
	7.0.0
	6.0.21
	6.0.20
	6.0.18
	6.0.16
	6.0.12
	6.0.11
	6.0.6
	6.0.5
	6.0.4
	6.0.3
	6.0.2
	6.0.1
	6.0.0
	5.0.17
	5.0.16
	5.0.15
	5.0.14
	5.0.13
	5.0.12
	5.0.11
	5.0.10
	5.0.9
	5.0.8
	5.0.7
	3.1.32
	3.1.31
	3.1.26
	3.1.25
	3.1.24
	3.1.23
	3.1.22
	3.1.21
	3.1.20
	3.1.19
	3.1.18
	3.1.17
	3.1.16

Xamarin.Profiler
Version : 1.8.0.49
Emplacement : /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version : 11

Outils de développement Apple
Xcode: 14.3.1 21815
Version : 14E300c

Xamarin.Android
Version : 13.2.1.2 (Visual Studio Community)
Commiter : xamarin-android/d17-5/a8a26c7
Android SDK: /Users/hamdani/Library/Developer/Xamarin/android-sdk-macosx
	Versions Android prises en charge :
		11.0(Niveau d’API 30)
		13.0(Niveau d’API 33)

Version des outils de ligne de commande du SDK : 7.0
Version des outils de plateforme du SDK : 34.0.1
Version des outils de génération du SDK : 32.0.0

Informations de build : 
Mono: d9a6e87
Java.Interop: xamarin/java.interop/d17-5@149d70fe
SQLite: xamarin/sqlite/3.40.1@68c69d8
Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-5@ca1552d

Microsoft Build of OpenJDK
Java SDK: /Library/Java/JavaVirtualMachines/microsoft-11.jdk
11.0.16.1
Android Designer code EPL disponible ici :
https://github.com/xamarin/AndroidDesigner.EPL

Eclipse Temurin JDK
Java SDK: /Library/Java/JavaVirtualMachines/temurin-8.jdk
1.8.0.302
Android Designer code EPL disponible ici :
https://github.com/xamarin/AndroidDesigner.EPL

Gestionnaire SDK Android
Version : 17.6.0.50
Synthèse : a715dca
Branche : HEAD
Date de build : 2023-08-02 19:04:28 UTC

Gestionnaire d'appareils Android
Version : 0.0.0.1309
Synthèse : 06e3e77
Branche : HEAD
Date de build : 2023-08-02 19:04:28 UTC

Xamarin Designer
Version: 17.6.3.9
Hash: 2648399ae8
Branch: remotes/origin/d17-6
Build date: 2023-08-02 19:04:23 UTC

Xamarin.Mac
Non installé

Xamarin.iOS
Version : 16.4.0.6 Visual Studio Community
Synthèse : 97731c92c
Branche : xcode14.3
Date de build : 2023-04-11 22:38:36-0400

Informations de build
Release ID: 1706030421
Git revision: c90350679f1b9295edc584519320b13d06489853
Build date: 2023-08-02 19:02:50+00
Build branch: release-17.6
Voie de build : release-17.6

Système d’exploitation
Mac OS X 13.4.1
Darwin 22.5.0 Darwin Kernel Version 22.5.0
    Thu Jun  8 22:22:22 PDT 2023
    root:xnu-8796.121.3~7/RELEASE_X86_64 x86_64

```

</details>

### Build Logs

[iOSBuildLog.zip](https://github.com/xamarin/xamarin-macios/files/12388221/iOSBuildLog.zip)

**Comment by rolfbjarne** [comment]
Does it work if you try a new project from a template?

If that fails, can you try a new project from a template directly from the Mac to see if that works?

**Comment by Bishokouts** [comment]
Hi @rolfbjarne , I created a new project from scratch with the visual studio mac blank template. I then managed to deploy the new APP on my iPhone successfully using the same automatic provisioning profile as on my other APP which does not work.

I tried to look at the differences between the iOS.csproj files but nothing significant, I did some tests without success.

**Comment by rolfbjarne** [comment]
What happens if you change the bundle identifier in the project that fails? This tells iOS it's a completely different app, so anything the iOS device might have cached will not apply anymore.

**Comment by Bishokouts** [comment]
I changed the bundle identifier, then cleaned the solution, recompiled and tried to deploy again. A new provisioning profile was automatically created by visual studio and the deployment attempt again failed with the same error message as before : 

VerifyingApplication - PercentComplete: 40%
 ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.uk83cY/extracted/[ProjectName].iOS.app : 0xe8008014 (The executable contains an invalid signature.)
error MT1006: Could not install the application '[Path of project]/bin/iPhoneSimulator/Debug/[ProjectName].iOS.app' on the device 'iPhone [My name]': AMDeviceSecureInstallApplicationBundle returned: 0xe8008014.

Is it normal that in the error message it indicates '[Path of project]/bin/**iPhoneSimulator**/Debug/[ProjectName].iOS.app' ?

**Comment by rolfbjarne** [comment]
> Is it normal that in the error message it indicates '[Path of project]/bin/**iPhoneSimulator**/Debug/[ProjectName].iOS.app' ?

No, that's not normal at all. That would certainly explain all the signing problems if the IDE for some reason is trying to install a simulator build on device.

Can you try install the app using the command line to see if that works?

```
/Library/Frameworks/Xamarin.iOS.framework/Versions/Current/bin/mlaunch -v -v -v -v --installdev /Users/hamdani/Projects/ChaskisDeliveryApp/ChaskisDeliveryApp.iOS/bin/iPhone/Debug/ChaskisDeliveryApp.iOS.app
```

if that doesn't work, please attach the output.

**Comment by tipa** [comment]
My error message looks a bit different, but I also try to deploy a "normal" NET7 iOS app, not a Forms app. This is the error the Console reports:
`0x16ba07000 -[MIClientConnection updatePlaceholderMetadataForApp:installType:failureReason:underlyingError:failureSource:completion:]: Update placeholder metadata requested by client installcoordinationd (pid 190 (501/501)) for app com.app installType = 7 failureReason = 14 underlyingError = Error Domain=MIInstallerErrorDomain Code=13 "Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.V9X5rC/extracted/my.iOS.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)" UserInfo={LibMISErrorNumber=-402620395, LegacyErrorString=ApplicationVerificationFailed, SourceFileLine=78, FunctionName=+[MICodeSigningVerifier _validateSignatureAndCopyInfoForURL:withOptions:error:], NSLocalizedDescription=Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.V9X5rC/extracted/my.iOS.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)} failureSource = <…>`

It reports a problem with the provisioning profile, but during build it appears like it found the correct one:
```
Detected signing identity:
        
  Code Signing Key: "Apple Development: Timo P (XXXXXX)" (xyzzy)
  Provisioning Profile: "timo dev" (guid)
  Bundle Id: com.app
  App Id: YYYYY.com.app
```

**Comment by rolfbjarne** [comment]
> It reports a problem with the provisioning profile, but during build it appears like it found the correct one:

Can you use the same provisioning profile in an Xcode project (with the same bundle identifier), using the same entitlements?

**Comment by tipa** [comment]
Ugh, I found the problem. In my case, I forgot to switch the `aps-environment` back to "development". Sorry for the false alarm

**Comment by Bishokouts** [comment]
@rolfbjarne I ran the command without success. 
Here is the output:

[output.txt](https://github.com/xamarin/xamarin-macios/files/12437151/output.txt)

**Comment by rolfbjarne** [comment]
> @rolfbjarne I ran the command without success. Here is the output:
> 
> [output.txt](https://github.com/xamarin/xamarin-macios/files/12437151/output.txt)

> ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.MWwel9/extracted/ChaskisDeliveryApp.iOS.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)

Are you using any custom entitlements in your app? It could be that the provisioning profile you're using doesn't allow some of those entitlements. If you are using any custom entitlements, can you try creating a dummy app in Xcode with the same bundle identifier and the same entitlements to see if that works? If so, afterwards try your project again, because Xcode will often auto-fix these types of issues.

**Comment by Bishokouts** [comment]
@rolfbjarne Yes I'm using a custom entitlements in my app.
I never had a provisioning file issue related to my entitlements before with my with the old version of my app which is currently online on the store with same entitlements.

So I creates a test APP in Xcode with same bundle ID and same entitlements (capabilities) as my APP. 

I used first "Automatic manage signing" in my Xcode test APP and It deployed successfully on my iPhone. When I tried to use same provisioning profil in my visual studio real APP I couldn't see Xcode provisioning profil. I think it was only a local provisioning profil.

So I created manually a provisioning profil on developer.apple.com I used this profil in my Xcode test APP and again successfully deployed on my iPhone. 
In my visual studio for mac real APP I used also manual provisioning with this exactly same profil but I got the same error as usual.
